components/motor/gpio/motor_encoder.go

```diff
@@ -354,184 +240,162 @@ func (m *EncodedMotor) rpmMonitor() {
 			continue
 		}
 		now := time.Now().UnixNano()
-		if now == lastTime {
-			// this really only happens in testing, b/c we decrease sleep, but nice defense anyway
+		atomic.AddInt64(&m.rpmMonitorCalls, 1)
+
+		if (m.DirectionMoving() == 1 && pos >= m.state.goalPos) || (m.DirectionMoving() == -1 && pos <= m.state.goalPos) {
+			// stop motor when at or past goal position
+			if err := m.Stop(m.cancelCtx, nil); err != nil {
+				m.logger.Error(err)
+				return
+			}
 			continue
 		}
-		atomic.AddInt64(&m.rpmMonitorCalls, 1)
-		// TODO: we round down here for absolute encoders, but absolute encoders
-		// should have their own logic separate from incremental
-		roundedPos := int64(math.Floor(pos))
-		inRamp = m.rpmMonitorPass(roundedPos, lastPos, now, lastTime, rpmDebug)
 
-		lastPos = int64(pos)
+		m.makeAdjustments(pos, lastPos, now, lastTime)
+
+		lastPos = pos
 		lastTime = now
 	}
 }
 
-// return is if we are in a ramp phase.
-func (m *EncodedMotor) rpmMonitorPass(pos, lastPos, now, lastTime int64, rpmDebug bool) bool {
+// makeAdjustments does the math required to see if the RPM is too high or too low,
+// and if the goal position has been reached.
+func (m *EncodedMotor) makeAdjustments(pos, lastPos float64, now, lastTime int64) {
 	m.stateMu.Lock()
 	defer m.stateMu.Unlock()
 
-	var ticksLeft int64
-
-	currentRPM := m.computeRPM(pos, lastPos, now, lastTime)
-	m.state.currentRPM = currentRPM
-
-	if !m.state.regulated && math.Abs(m.state.desiredRPM) > 0.001 {
-		m.rpmMonitorPassSetRpmInLock(currentRPM, m.state.desiredRPM, -1, rpmDebug)
-		return false
+	// calculate RPM based on change in position and change in time
+	deltaPos := (pos - lastPos) / float64(m.ticksPerRotation)
+	deltaTime := (float64(now) - float64(lastTime)) / float64(6e10)
+	if deltaTime == 0.0 {
+		m.state.currentRPM = 0
+	} else {
+		m.state.currentRPM = deltaPos / deltaTime
 	}
 
-	if !m.state.regulated {
-		return false
-	}
-
-	// correctly set the ticksLeft accounting for power supplied to the motor and the expected direction of the motor
-	ticksLeft = (m.state.setPoint - pos) * sign(m.state.lastPowerPct) * m.flip
-	rotationsLeft := float64(ticksLeft) / float64(m.ticksPerRotation)
+	dir := m.directionMovingInLock()
 
-	if rotationsLeft <= 0 { // if we have reached goal or overshot, turn off
-		if rpmDebug {
-			m.logger.Debugf("rot %.2f, stopping motor", rotationsLeft)
+	// "controls"
+	quartRot := float64(m.ticksPerRotation) / 4.0
+	if pos < m.state.goalPos+quartRot && pos > m.state.goalPos-quartRot {
+		powerPct := m.state.lastPowerPct - (m.rampRate * 2 * m.directionMovingInLock())
+		if sign(powerPct) != m.directionMovingInLock() {
+			powerPct *= -1
 		}
-		err := m.off(m.cancelCtx)
-		if err != nil {
-			m.logger.Warnf("error turning motor off from after hit set point: %v", err)
+		m.state.lastPowerPct = powerPct
+		if err := m.setPower(m.cancelCtx, powerPct, true); err != nil {
+			panic(err)
```

Comment:
maybe log the error and return/exit from this function and the loop.

Reply:
unsure how to return from the loop. doesn't panic terminate everything?

Reply:
Then let's return an error from this function and check it in the main loop. panic does terminate everything, but that's why it's dangerous - what is it leaving behind? how deterministically is it handling keeping the state of the power in the adjustemnt.